File: Week 3/BOJ_실버2_나무 자르기/hyungjoon_failed2.java
import java.util.*;
import java.io.*;

/**
 * 문제 : 나무 자르기
 * 난이도 : 실버 2
 * 소요 시간 : 80분
 */
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        int[] arr = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();

        int answer = 0;
        // 나무들은 정렬되어 있지 않다. 따라서, 나무를 대상으로 l과 r포인터를 잡는게 아니라,
        // 높이를 기준으로 l과 r포인터를 잡아야 한다.
        int l = 0, r = Arrays.stream(arr).max().getAsInt();
        while(l <= r){
            // 기준을 잡을 mid 값
            int mid = (l+r) / 2;
            int temp = 0;
            for(int i = 0; i < n; i++){
                if(arr[i] > mid){
                    temp += arr[i] - mid;
                }
            }

            // 벌목량이 m 이상이면 정답 갱신
            if(temp >= m){
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        System.out.println(l+1);
    }
}

File: Week 3/BOJ_골드5_수 고르기/hyungjoon_failed2.py
'''
 문제 : 수 고르기
 난이도 : 골드 5
 소요 시간 : 
'''
import sys
se = sys.stdin.readline

n, m = map(int, se().split())
arr = [ int(se().rstrip()) for _ in range(n) ]

# 정렬시키고 투포인터 진행
arr.sort()

l, r = 0, 0
answer = int(2e9) # 무한대, 10억
while(l <= r and r < n):
    temp = arr[r] - arr[l]
    # r과 l의 차이가 m 이상인 경우는 r만 계속 줄여주면 된다.
    if temp >= m:
        answer = min(answer, temp)
        l += 1
        # r이 n-1보다 작고, l과 r의 차이가 1밖에 안나는 상황이면 추가적인 탐색을 해야하니 r을 늘려준다.
        if r - l == 1 and r < n-1:
            r += 1
    else:
        r += 1
print(answer)